protocol/messenger_unread_test.go

```diff
@@ -0,0 +1,277 @@
+package protocol
+
+import (
+	"context"
+	"time"
+
+	_ "github.com/mutecomm/go-sqlcipher/v4" // require go-sqlcipher that overrides default implementation
+
+	"github.com/status-im/status-go/protocol/common"
+)
+
+func (s *MessengerSuite) TestMarkMessageAsUnreadWhenMessageListContainsSingleMessage() {
+	chat := CreatePublicChat("test-chat-1", s.m.transport)
+
+	chat.UnviewedMessagesCount = 2
+	chat.UnviewedMentionsCount = 2
+	chat.Highlight = true
+
+	err := s.m.SaveChat(chat)
+	s.Require().NoError(err)
+
+	inputMessage1 := buildTestMessage(*chat)
+	inputMessage1.ID = "1"
+	inputMessage1.Seen = true
+	inputMessage1.Mentioned = true
+
+	err = s.m.SaveMessages([]*common.Message{inputMessage1})
+	s.Require().NoError(err)
+
+	_, err = s.m.MarkAllRead(context.Background(), chat.ID)
+	s.Require().NoError(err)
+
+	chats := s.m.allChats
+
+	actualChat, ok := chats.Load(chat.ID)
+
+	s.Require().Equal(true, ok)
+	s.Require().Equal(uint(0), actualChat.UnviewedMessagesCount)
+	s.Require().Equal(uint(0), actualChat.UnviewedMentionsCount)
+	s.Require().Equal(false, actualChat.Highlight)
+
+	actualMessage, err := s.m.MessageByID(inputMessage1.ID)
+	s.Require().NoError(err)
+	s.Require().Equal(true, actualMessage.Seen)
+
+	count, countWithMentions, err := s.m.MarkMessageAsUnread(chat.ID, inputMessage1.ID)
+	s.Require().NoError(err)
+	s.Require().Equal(uint64(1), count)
+	s.Require().Equal(uint64(1), countWithMentions)
+
+	chats = s.m.allChats
+
+	actualChat, ok = chats.Load(chat.ID)
+	s.Require().Equal(true, ok)
+	s.Require().Equal(uint(1), actualChat.UnviewedMessagesCount)
+	s.Require().Equal(uint(1), actualChat.UnviewedMentionsCount)
+	s.Require().Equal(false, actualChat.Highlight)
+
+	actualMessage, err = s.m.MessageByID(inputMessage1.ID)
+	s.Require().NoError(err)
+	s.Require().Equal(false, actualMessage.Seen)
+}
+
+func (s *MessengerSuite) TestMarkMessageAsUnreadWhenMessageListContainsSeveralMessages() {
+	chat := CreatePublicChat("test-chat-2", s.m.transport)
+
+	chat.UnviewedMessagesCount = 0
+	chat.UnviewedMentionsCount = 0
+	chat.Highlight = true
+
+	err := s.m.SaveChat(chat)
+	s.Require().NoError(err)
+
+	inputMessage1 := buildTestMessage(*chat)
+	inputMessage1.ID = "1"
+	inputMessage1.Seen = true
+	inputMessage1.Mentioned = true
+
+	time.Sleep(1000 * time.Millisecond)
```

Comment:
still leftover? 🙂

Reply:
I used it to simulate the timestamp as the timestamp is created automatically when creating the message and to account for the speed of unittest, but I understand I can directly overid the Timestamp to make the test more controllable here. I'll update this. thanks